spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapper.java

```diff
@@ -0,0 +1,83 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.HeaderMapper;
+import org.springframework.util.Assert;
+
+public class PubSubHeaderMapper implements HeaderMapper<Map<String, String>> {
+
+	/**
+	 * Headers to exclude in the {@link #fromHeaders(MessageHeaders, Map)}.
+	 */
+	private String[] doNotMapHeaderNames = {
+			MessageHeaders.ID, MessageHeaders.TIMESTAMP
+	};
+
+	private boolean filterHeaders;
+
+	/**
+	 * Set the names of the headers to filtered out in {@link #fromHeaders(MessageHeaders, Map)}.
+	 * @param doNotMapHeaderNames
+	 */
+	public void setDoNotMapHeaderNames(String... doNotMapHeaderNames) {
+		Assert.notNull(doNotMapHeaderNames, "Header names can't be null.");
+		Assert.noNullElements(doNotMapHeaderNames, "No header name can be null.");
+		this.doNotMapHeaderNames = doNotMapHeaderNames;
+	}
+
+	/**
+	 * Generate headers in {@link com.google.pubsub.v1.PubsubMessage} format from
+	 * {@link MessageHeaders}. All headers are converted into strings.
+	 *
+	 * <p>May filter out designated headers, depending on whether {@code filterHeaders} is
+	 * {@code true}.
+	 * @param messageHeaders headers to map from
+	 * @param pubsubMessageHeaders headers in their final format
+	 */
+	@Override
+	public void fromHeaders(MessageHeaders messageHeaders,
+			final Map<String, String> pubsubMessageHeaders) {
+		messageHeaders.forEach((key, value) -> pubsubMessageHeaders.put(key, value.toString()));
+
+		if (this.filterHeaders) {
+			Arrays.stream(this.doNotMapHeaderNames).forEach(messageHeaders::remove);
+		}
+	}
+
+	/**
+	 * Generate headers in {@link org.springframework.messaging.Message} format from
+	 * {@code Map<String, String>}.
+	 * @param pubsubMessageHeaders headers in {@link com.google.pubsub.v1.PubsubMessage} format
+	 * @return a map with headers in the {@link org.springframework.messaging.Message} format
+	 */
+	@Override
+	public MessageHeaders toHeaders(Map<String, String> pubsubMessageHeaders) {
+		Map<String, Object> typifiedHeaders = new HashMap<>();
+		pubsubMessageHeaders.forEach((key, value) -> typifiedHeaders.put(key, value));
```

Comment:
Why do you map all headers from here?
Shouldn't we filter them as well?

Shouldn't we have a separate set of headers to map `to` and `from`?
See, `StompHeaderMapper` for example: https://github.com/spring-projects/spring-integration/blob/master/spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java

Reply:
I can filter the maps as I add them, if you want.
If you notice, first I add all the headers, then I remove the ones to be filtered out.
Since it is a map, accesses are O(1), so I thought it would be fine, but more readable like this.
I can change it if you like, though.

Pub/Sub doesn't really add any headers to the messages and I'm not sure a set of `toHeaders` headers to filter would be useful. YAGNI?

Reply:
???
This is  `toHeaders()` method and I fully don't see any filtering logic here, therefore your explanation just does not make sense to me.

Please, try to express your mind one more time.

Sorry for confusing you, but I can't accept what I don't understand.

Reply:
Sorry, I realised that late and edited my comment.
I'm not sure if it's worth to filter headers on `PubsubMessage`s because Pub/Sub doesn't add any headers. It feels like a feature that isn't necessary?